src/core/pininput.cpp

```diff
@@ -2094,3 +2149,605 @@ LPDIRECTINPUTDEVICE PinInput::GetJoystick(int index)
+   uint8_t lrFlipper; // lower right flipper button duty cycle
+   uint8_t ulFlipper; // upper left flipper button duty cycle
+   uint8_t urFlipper; // upper right flipper button duty cycle
+   int16_t axNudge; // instantaneous nudge acceleration, X axis (left/right)
+   int16_t ayNudge; // instantaneous nudge acceleration, Y axis (front/back)
+   int16_t vxNudge; // instantaneous nudge velocity, X axis
+   int16_t vyNudge; // instantaneous nudge velocity, Y axis
+   int16_t plungerPos; // current plunger position
+   int16_t plungerSpeed; // instantaneous plunger speed
+
+   // load the struct from the USB byte format
+   void LoadFromUSB(const uint8_t *p, size_t reportSize)
+   {
+      // clear all fields
+      memset(this, 0, sizeof(*this));
+
+      // Load fields until we exhaust the data.  We can stop as soon as we
+      // run out of input bytes, since we've already cleared all of the 
+      // elements.
+      Load(timestamp, p, reportSize)
+      && Load(genericButtons, p, reportSize)
+      && Load(pinballButtons, p, reportSize)
+      && Load(llFlipper, p, reportSize)
+      && Load(lrFlipper, p, reportSize)
+      && Load(ulFlipper, p, reportSize)
+      && Load(urFlipper, p, reportSize)
+      && Load(axNudge, p, reportSize)
+      && Load(ayNudge, p, reportSize)
+      && Load(vxNudge, p, reportSize)
+      && Load(vyNudge, p, reportSize)
+      && Load(plungerPos, p, reportSize)
+      && Load(plungerSpeed, p, reportSize);
+   }
+
+   // Load bytes from the USB byte format into a local native integer type.
+   // This correctly translates from the packed little-endian wire format
+   // into the local integer representation.  Note: ele must be pre-cleared
+   // to zero before the call.
+   template <typename T> static bool Load(T &ele, const uint8_t* &p, size_t &sizeRemaining)
+   {
+      // figure how many bytes we have to read for the complete type
+      size_t nBytesInT = sizeof(T) / sizeof(uint8_t);
+      if (sizeRemaining >= nBytesInT)
+      {
+         // we can complete the full read
+         for (size_t i = 0; i < nBytesInT ; ++i, ele |= static_cast<T>(*p++) << (i*8));
+
+         // deduct the size consumed and return success
+         sizeRemaining -= nBytesInT;
+         return true;
+      }
+      else
+      {
+         // we can't complete the read, so we've exhausted the input - clear
+         // all remaining bytes (so that we don't try to interpret the remaining
+         // bytes as a smaller type that occurs later in the struct) and return 
+         // end-of-file
+         sizeRemaining = 0;
+         return false;
+      }
+   }
+};
+
+// the USB implementation is in terms of the hidapi library
+#include "third-party/include/hidapi/hidapi.h"
+
+// NON-PORTABLE FUNCTION INTERFACE: Inspect a hidapi device's HID Input
+// Report Descriptor to determine if this is an Open Pinball Device.
+// hidapi doesn't provide access to the Report Descriptors, so we have

```

Comment:
Is there an issue on the side of hidapi to track this missing feature? Might be a good idea to link it here.

Reply:
I doubt it - I think it's just outside the scope of what hidapi sets out to do.

Reply:
There's a long conversation at https://github.com/libusb/hidapi/issues/249

Reply:
It should be exactly what's needed, and it looks like the new facilities are in the library version that Jason's, so I'll look into switching over to that. If it works it should eliminate the extra platform-specific function.